### ISSUE TITLE
Weird behavior with broadcast streams

### ISSUE THREAD

**Issue by shinayser** [ReactiveX/rxdart#308]
This thing was killing me until a found whats going on. I will try to explain it by code:

First, check this stream transformer:
``` dart
StreamTransformer<int, int> emitsAfter5() {
  return StreamTransformer<int, int>((Stream<int> stream, bool cancelOnError) {
    StreamController<int> controller;
    StreamSubscription<int> subscription;

    var amount = 0;

    controller = StreamController<int>(
        sync: true,
        onListen: () {
          subscription = stream.listen((value) {
            print("RECEIVED $value");
            try {
              if ((++amount) == 5) {
                controller.add(value);
                amount = 0;
              }
            } catch (e, s) {
              controller.addError(e, s);
            }
          },
              onError: controller.addError,
              onDone: controller.close,
              cancelOnError: cancelOnError);
        },
        onPause: ([Future<dynamic> resumeSignal]) =>
            subscription.pause(resumeSignal),
        onResume: () => subscription.resume(),
        onCancel: () => subscription.cancel());

    return controller.stream.listen(null);
  });
}
```

It simply prints "RECEIVED" when a new value comes to it but only forwards it when its the 5° value received. 

Now to test it you can just:

``` dart
void main() async {
  var stream = Stream.periodic(seconds(1), (index) => index);

  var observable = Observable(stream).transform(emitsAfter5());

  observable.listen((it) => print("Arrived on 1: $it"));
  observable.listen((it) => print("Arrived on 2: $it"));
}
```

You will se the following error:
```
Bad state: Stream has already been listened to.
```
Perfectly normal, since the "Stream.periodic" is not an BroadcastStream. So lets fix it, an now things are going to get weird:

``` dart
void main() async {
  var stream = Stream.periodic(seconds(1), (index) => index).asBroadcastStream();

  var observable = Observable(stream).transform(emitsAfter5());

  observable.listen((it) => print("Arrived on 1: $it"));
  observable.listen((it) => print("Arrived on 2: $it"));
}
```

Will print:
```
RECEIVED 0
RECEIVED 0
RECEIVED 1
RECEIVED 1
RECEIVED 2
RECEIVED 2
RECEIVED 3
RECEIVED 3
```

Wait? Why is it printing twice? There is only one Stream being subscribed! After hours trying to figure out i found the sollution: adding `asBroadcastStream()` on the OSBERVABLE too!

``` dart
void main() async {
  var stream =
      Stream.periodic(seconds(1), (index) => index).asBroadcastStream();

  var observable =
      Observable(stream).transform(emitsAfter5()).asBroadcastStream();

  observable.listen((it) => print("Arrived on 1: $it"));
  observable.listen((it) => print("Arrived on 2: $it"));
}
```
Results:
```dart
RECEIVED 0
RECEIVED 1
RECEIVED 2
RECEIVED 3
RECEIVED 4
```
Now everything works.

So the question is? Is this a BUG? Having to set "asBroadcastStream" twice (one for the stream and one for the observable) is the correct way to go? That sounds weird to me and pehaps are causing some extra, uncessessary processing on apps out there in the wild.

**Comment by frankpepermans** [comment]
Hi there,

This is Dart Streams behavior, not rxdart,
I'm on a phone atm, so short answer:

Check the transformer you wrote, the print happens inside the onListen method.

This literally means that this method triggers on each listen you apply, so having 2x output with 2x listen is as expected.

The asBroadcastStream trick, I'd need to check on an actual computer to test, for now, could you remove any rxdart from your code and try it out with regular Streams? You should get the same print outputs.

**Comment by shinayser** [comment]
> Hi there,
> 
> This is Dart Streams behavior, not rxdart,
> I'm on a phone atm, so short answer:
> 
> Check the transformer you wrote, the print happens inside the onListen method.
> 
> This literally means that this method triggers on each listen you apply, so having 2x output with 2x listen is as expected.
> 
> The asBroadcastStream trick, I'd need to check on an actual computer to test, for now, could you remove any rxdart from your code and try it out with regular Streams? You should get the same print outputs.

Hey frank! So, let me clarify:
1. The print happens not on the "listen" but on the "onData". So, everytime a new data arrives, it prints.

2. I did the test you asked as follow:

```dart
void main() async {
  var stream = Stream.periodic(seconds(1), (index) => index)
      .transform(emitsAfter5())
      .asBroadcastStream();

  stream.listen((it) => print("Arrived on 1: $it"));
  stream.listen((it) => print("Arrived on 2: $it"));
}
```

The result was:
```
RECEIVED 0
RECEIVED 1
RECEIVED 2
RECEIVED 3
RECEIVED 4
```

As you can see, it worked! Wich makes me wonder that the Observable is not taking the "asBroadcastStream" from the input `Stream` in consideration.

**Comment by shinayser** [comment]
**I am reworking my sample and replacing my StreamTransformer with a simple `map`to makes things clearer:**

This first code works properly:

```dart
void main() async {
  var stream = Stream.periodic(ms(500), (it) => it).map((it) {
    print("Calling this inside of 'map' with value $it");
    return it;
  }).asBroadcastStream();

  stream.listen((it) {
    //Do nothing
  });

  stream.listen((it) {
    //Do nothing again
  });
}

```

This second code, shows us the duplicated outputs using Observables:

```dart
void main() async {
  var stream = Stream.periodic(ms(500), (it) => it).asBroadcastStream();

  var observable = Observable(stream).map((it) {
    print("Calling this inside of 'map' with value $it");
    return it;
  });

  observable.listen((it) {
    //Do nothing
  });

  observable.listen((it) {
    //Do nothing again
  });
}
```

And finally, this third code fixes the second code above using "asBroadcastStream" on the Observable:

```dart
void main() async {
  var stream = Stream.periodic(ms(500), (it) => it).asBroadcastStream();

  var observable = Observable(stream).map((it) {
    print("Calling this inside of 'map' with value $it");
    return it;
  }).asBroadcastStream();

  observable.listen((it) {
    //Do nothing
  });

  observable.listen((it) {
    //Do nothing again
  });
}
```

### So, what should be the correct behavior?

In my opinion, what is going wrong here is that the `Observable` is not taking the "asBroadcastStream" coming from the `Stream` in consideration, making us to add another "asBroadcastStream" manually on the `Observable`.

It should be detected automatically:

1. Observable constructor should look the **isBroadcast** property from the incoming stream.
2. If it is **true**, it turns on his own broadcasting behavior.

**Comment by frankpepermans** [comment]
Will check,

Might be related to this: https://github.com/dart-lang/sdk/issues/36965

**Comment by shinayser** [comment]
> Will check,
> 
> Might be related to this: [dart-lang/sdk#36965](https://github.com/dart-lang/sdk/issues/36965)

Yes! This is probably the reason. Check this code, it also gives duplicated prints even after calling asBroadcastStream before the map:

```dart
void main() async {
  var stream = Stream.periodic(ms(500), (it) => it).asBroadcastStream();

  var observable = Observable(stream)
      .asBroadcastStream() // <<-- Now, calling the asBroadcastStream before the map
      .map((it) { 
    print("Calling this inside of 'map' with value $it");
    return it;
  });

  observable.listen((it) {
    //Do nothing
  });

  observable.listen((it) {
    //Do nothing again
  });
}
```

So, the problem is not with rxdart but with Dart streams itself?

**Comment by frankpepermans** [comment]
Yup, I'll try to have another look when I've got some time, hopefully the Dart team will fix it soon :)

**Comment by shinayser** [comment]
Since this is not a rxdart issue, I am closing this issue. If someone else wants to know the progress of this fix, just follow dart-lang/sdk#36965.
